assertsprocessor/sampler.go

```diff
@@ -0,0 +1,195 @@
+		}
+	}
+}
+
+func (s *sampler) getSummary(traceId string, spanSet *resourceSpanGroup) *traceSummary {
+	summary := traceSummary{}
+	summary.hasError = false
+	summary.isSlow = false
+	maxLatency := float64(0)
+	summary.hasError = summary.hasError || spanSet.hasError()
+	entityKey := buildEntityKey(s.config, spanSet.namespace, spanSet.service)
+	for _, rootSpan := range spanSet.rootSpans {
+		request := getRequest(s.requestRegexps, rootSpan)
+		summary.isSlow = summary.isSlow || s.isSlow(spanSet.namespace, spanSet.service, rootSpan)
+		max := math.Max(maxLatency, computeLatency(rootSpan))
+		if max > maxLatency {
+			maxLatency = max
+			summary.slowestRootSpan = rootSpan
+			summary.requestKey = RequestKey{
+				entityKey: entityKey,
+				request:   request,
+			}
+			summary.latency = maxLatency
+		}
+	}
+	s.logger.Info("Trace summary",
+		zap.String("traceId", traceId),
+		zap.String("request", summary.requestKey.AsString()),
+		zap.Bool("slow", summary.isSlow),
+		zap.Bool("error", summary.hasError),
+		zap.Float64("latency", summary.latency))
+	return &summary
+}
+
+func (s *sampler) getTraceQueues(key RequestKey) *traceQueues {
+	var pq, _ = s.topTracesMap.LoadOrStore(key.AsString(), &traceQueues{
+		slowQueue:  NewTraceQueue(int(math.Min(5, float64(s.config.MaxTracesPerMinutePerContainer)))),
+		errorQueue: NewTraceQueue(int(math.Min(5, float64(s.config.MaxTracesPerMinutePerContainer)))),
+	})
+	return pq.(*traceQueues)
+}
+
+func (s *sampler) isSlow(namespace string, serviceName string, rootSpan ptrace.Span) bool {
+	spanDuration := computeLatency(rootSpan)
+	threshold := s.thresholdHelper.getThreshold(namespace, serviceName, rootSpan.Name())
+	return spanDuration > threshold
+}
+
+func (s *sampler) stopFlushing() {
+	func() { s.stop <- true }()
+}
+
+func (s *sampler) flushTraces() {
+	for {
+		select {
+		case <-s.stop:
+			s.logger.Info("Trace flush background routine stopped")
+			return
+		case <-s.traceFlushTicker.C:
+			var previousTraces = s.topTracesMap
+			s.topTracesMap = &sync.Map{}
+			var serviceQueues = map[string]*TraceQueue{}
+			previousTraces.Range(func(key any, value any) bool {
+				var requestKey = key.(string)
+				var scopeString, part1, _ = strings.Cut(requestKey, "#")
+				var typeName, part2, _ = strings.Cut(part1, "#")
+				var name, _, _ = strings.Cut(part2, "#")
+				var entityKey = scopeString + "#" + typeName + "#" + name
+				var q = value.(*traceQueues)
+
+				// Flush all the errors
+				if len(q.errorQueue.priorityQueue) > 0 {
+					s.logger.Info("Flushing Error Traces for",
+						zap.String("Request", requestKey),
+						zap.Int("Count", len(q.errorQueue.priorityQueue)))
+					for _, item := range q.errorQueue.priorityQueue {
+						_ = (*s).nextConsumer.ConsumeTraces(*item.ctx, *item.trace)
+					}
+				}
+
+				if len(q.slowQueue.priorityQueue) > 0 {
+					serviceQueue := serviceQueues[entityKey]
+					if serviceQueue == nil {
+						serviceQueue = NewTraceQueue(s.config.MaxTracesPerMinute)
```

Comment:
This is a bit late to be enforcing the overall limit. We should limit it while adding it to the priority queues itself